### PR TITLE
fix(ui): Restrict visibility of each component/release like Project

### PR DESCRIPTION
**Pull request by KoukiHama**
Signed-off-by: Kouki Hama <kouki1.hama@toshiba.co.jp>

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> * Which issue is this pull request belonging to and how is it solving it? (*Refer to issue here*)
> * Did you add or update any new dependencies that are required for your change?

Issue: #1155

### How To Test?
Listed at the bottom in this comment.

# The outline of a functional improvement
## Purpose
This enables users to change the disclosure range of Component by setting of Visibility, as well as the Project.
To adjust this fix, add the property which holds a business unit in Component.

## The outline of function change

1. Add a visibility to the property of Component.
     - Component detail view
     - Component edit view
     - Component duplication view

1. Add a business unit to the property of Component.
     - Component detail view
     - Component edit view
     - Component duplication view

1.  Correspondence to Restrictions on Access  
     Based on a user's browsing permission, the display of Components, and the related Releases are restricted.
     - Component section (Component list display)
    Components
     - Component detail view  
          - Release Overview subsection (list display of Release)
          - Summary subsection (List display of Component in use)
         Component_Detail_Summary_Using-Component
          - Restrict access without the authority by URL.
     - Component edit view  
          - Releases subsection (list display of Release)
          - Release selection dialog (selection of Release)
          - Restrict access without the authority by URL.
     - Release detail view  
          - Linked releases subsection (list display of Release)
Release_Detail_Linked-releases
          - Summary subsection (List display of Component in use)
         Release_Detail_Summary_Using-Component
          - Restrict access without the authority by URL.
     - Release edit view  
          - Linked releases subsection (list display of Release)
         Release_Edit_Linked-release
          - Restrict access without the authority by URL.
     - Release duplication view  
          - Linked releases subsection (list display of Release)
          - Restrict access without the authority by URL.
     - Project detail view
          - License Clearing subsection (display of linked Release)
          Project_Detail_License-Clearing
         Project_Detail_License-Clearing_ListView
          Generate license information  
          Project_Detail_License-Clearing_Generate-Licence-Info_fixed
          Generate source cod bundle  
         Project_Detail_License-Clearing_Generate-Source-Code-Bundle
          - ECC subsection (display of linked Release)
         Project_Detail_ECC
          - Attachment Usages subsection (display of linked Release)  
          Project_Detail_Attachment-Usage

     - Project edit view
          - Linked Releases And Projects subsection (display of linked Release)         
         Project_Edit_Linked-Releases-And-Projects
          - Release selection dialog (Selection of Release)
     - Project duplication view
          - Linked Releases And Projects subsection (display of linked Release)  
          - Release selection dialog (Selection of Release)
     - Home section
          - RECENT COMPONENTS
          - RECENT RELEASES
         Home_Recent-components_Recent-releases

     - ECC section (display of Release)
     - Search section (display search results of Component and Release)
     - Correction of the error message of a circular reference links.
     Release_Edit_CyclicLinkError

****
1.  Enable users to specify Group (business unit) by Advanced Search of Component.
    Components_AdvancedSearch

1.  Restrict the use of the Merge function of Component and the Split function, other than Admin.
1.  Correspondence to Component Addition Property of CSV Import Function  
     Set these following values to visibility and businessUnit which were added to the Component.  
     visibility = EVERYONE  
     businessUnit = Primary department name of the user who performed the CSV import.  

1.  Correction of Export Spreadsheet Function of Project and Component  
     Release relevant to Component without READ authority is displayed as InaccessibleRelease.

## Points of code change
1.  Front End
     - The realization method of an access improper display  
     When acquiring Component and Release information from a back end server, authority information is given to PermissionMap of a data object.  
     the Front end program displays the object which does not have read permission as an Inaccessible object,  in reference to the Read permission of PermissionMap.

1.  Back End  
     - About naming of Thrift API of Component  
     Changes to the API on the backend side regarding Component's browsing restrictions.  
     The existing APIs are not changed, and either of the following two types of extended APIs are added as needed.
     
          - Accessible***  
          Example) GetComponents -> GetAccessibleComponents  
          In an API that returns a list of components, only those that can be viewed by the requesting User are returned.  
     
          - ***WithAccessibility       
          Example) GetComponents -> GetComponentsWithAccessibility  
          In an API that returns a list of components, the returned item is the same as in the original API. The Back end service decides the actions that the requesting User is authorized to perform on the Component, and attach the Permission Map to the result according to it.

     - Permission judging  
     Control the READ action to Component based on the visibility setting.  
     No changes will be made to permission control except for the READ action.  

          |Action|Admin|Clearing Admin|Moderators|Contributors|Creator|Same Group Users|Other Users|
          |-|-|-|-|-|-|-|-|
          |READ - PRIVATE|X|-|-|-|X|-|-|
          |READ - ME_AND_MODERATORS|X|-|X|X|X|-|-|
          |READ - BUISNESSUNIT_AND_MODERATORS|X|X|X|X|X|X|-|
          |READ - EVERYONE|X|X|X|X|X|X|X|
          |WRITE (not changed)|X|X|X|X|X|-|-|
          |ATTACHMENTS (not changed)|X|X|X|X|X|-|-|
          |DELETE (not changed)|X|-|X|X|X|-|-|
          |USERS (not changed)|X|-|X|X|X|-|-|
          |CLEARING (not changed)|X|-|X|X|X|-|-|
          |WRITE_ECC (not changed)|X|-|-|-|-|-|-|

          X : Access permission  
          \- : Access denied  

     - Read action of Release  
     Allow access to Release if the READ action can be performed on the Component holding the Release.  
     Since the read permissions of Release cannot be determined only by Release, they are controlled by determining the permissions of Release each time, not by ReleasePermissions.  

## Change of data structure
- The additional variable to Component
     - optional Visibility visibility(default=EVERYONE)  
          The disclosure range of Component setting.
          The Visibility type is as same as the disclosure range setting of Project.
     - optional string businessUnit  
          The organization that manages Component.  
- The additional variable to ReleaseLink  
     - optional bool accessible = true  
          The Flag to hold the accessibility of the linked Release.  
- ReleaseClearingStatusData  
     - optional bool accessible = true  
          The Flag to hold the accessibility of the linked Release.  

## Migration of DB
DB migration scripts provided, and added properties to Component.
A migration script performs the following operations to the Component type data in Couchdb.  
1.  When visibility is not set for the Component, set EVERYONE.
1.  When businessUnit is not set for the Component, set the Department of the primary of createdBy user.  

## How to test
1.  Advance Preparations
     - Create Componet which belongs to DEPARTMENT, DEPARTMENT2, and DEPARTMENT3, respectively.
     - Register Release into all Components created above.
1.  Check Additional Property
     - Select a Group in the Component edit view, and check the Group set in the Component detail view.  
     - Select Visibility in the Component edit view, and check the visibility which was set in the Component detail view.  
1.  Check Access Permission  
     - Check the viewing that restricted by the Visibility of the Component and the attribute of the logged-in user, according to the following table.

          |Action|Admin|Clearing Admin|Moderators|Contributors|Creator|Same Group Users|Other Users|
          |-|-|-|-|-|-|-|-|
          |READ - PRIVATE|X|-|-|-|X|-|-|
          |READ - ME_AND_MODERATORS|X|-|X|X|X|-|-|
          |READ - BUISNESSUNIT_AND_MODERATORS|X|X|X|X|X|X|-|
          |READ - EVERYONE|X|X|X|X|X|X|X|

          X : Access permission  
          \- : Access denied  

     - Check that Components and Releases without access permission are not displayed in the "Correspondence to Restrictions on Access" definition screen which is described above.  

1.  Operation Verification of Migration Script
     - The script for migrations  
          047_add_component_businessunit.py
     - Test procedure
          1.  Open a script by a text editor and change the following parameters.
               - DRY_RUN  
               Change the effected invalidity of Dryrun (test mode which does not perform the writing to DB)
                    - True:   The writing to DB is not performed.
                    - False:   The writing to DB is performed.
               - COUCHSERVER  
                    The host address of couchdb
               - DB_USER_NAME  
                    The administrator's name needed for the certification of couchdb.
                    When this was blank, the credential is not set up.
                    Unnecessary, when administrator account is not set to DB.
               - DB_USER_PASSWORD  
                    The administrator password of couchdb
          1.  Start the system before merging.
          1.  Login as user User1 who belongs to DEPARTMENT, and create component MigrationTestComponent001.
          1.  Login as user User2 who belongs to DEPARTMENT2, and create component MigrationTestComponent002.
          1.  Shut down the system.
          1.  Start couchdb.
          1.  Perform a migration script.
          1.  Start the system after merging.
          1.  Log in by an Admin user and check that Group and Visibility of MigrationTestComponent001 and MigrationTestComponent002 are set as the following values.  
               - MigrationTestComponent001
                    - Group: DEPARTMENT
                    - Visibility: EVERYONE
               - MigrationTestComponent002
                    - Group: DEPARTMENT2
                    - Visibility: EVERYONE

## limitations

1.  Property Names of Visibility  
Component's disclosure range property is a reuse of Project's code and uses the same property name as Project.
For this reason, the same problem as following issue exists.
There are many "Visbility" in source code. Are these typo?Should we rewrite to "Visibility" them for readability?#1311

### CODE REVIEW

**Comment by KoukiHama** [comment]
I solved merge confilct.

**Comment by KoukiHama** [comment]
I solved merge conflicts again and cleaned up commit logs.

**Review by ravi110336 — COMMENTED** [review]
Reviewed and tested the use case please consider some review comments and see some observations. 
1. In the component edit view page, can we shift the **Group** to after **Moderators** or may be in the beginning like it is there for the Edit Project view, so that we can remove this additional row in the UI.
**Component Edit View:**
component Group
**Project Edit View:**
ProjectEditView

2. Currently in the Project view page in edit mode the Group is editable in the text field and it takes the user organization value from the Liferay DB and from the couchDB, but in the Component edit view the Group is a drop down list. Could you please confirm if this feature for the component view is correct, since it differs from the Project view for the user group.

component Group


3. Private component/release is showing Inaccessible component/release for the other users,so if we have more private component then all will show inaccessible in the UI for other users. But currently in the Projects view,if we set a project visibilty to private then it won't be visible to the other users, but in the new changes for component/releases it shows the text as "Inaccessible component"/"Inaccessible release". is this correct?

4. For other user where the release is Inaccessible for him due to the visibility as Private.
Go to Edit Component->Open a release->Edit Release->linked releases
it shows the Inaccessible release for the user. 
Now, if we try to add/link a new release. Here the **Inaccessible release**  text looks bit odd as per the visibility in the UI.
linked releases




5. I am not able to see the below highlighted view (Summary subsection (List display of Component in use)), Please let me know if i missing the step to produce this.
components

**Comment by KoukiHama** [comment]
@ravi110336 Thank you for checking and giving feedback about  this pull request!

I will answer  all your question and fix codes following your comments soon,

Before, I start to update this pull request, I want you check the new design about linked release page.

> For other user where the release is Inaccessible for him due to the visibility as Private.
Go to Edit Component->Open a release->Edit Release->linked releases
it shows the Inaccessible release for the user.
Now, if we try to add/link a new release. Here the Inaccessible release text looks bit odd as per the visibility in the UI.

I considered new one base on your comment, and this is new one. What do you think ? thank you!

image

**Comment by ravi110336** [comment]
Hi @KoukiHama , Thanks for the update. The above design looks good to me, as discussed in the call we can change the text as "Restricted component/release" instead of Inaccessible.

**Comment by KoukiHama** [comment]
> 1. In the component edit view page, can we shift the Group to after Moderators or may be in the beginning like it is there for the Edit Project view, so that we can remove this additional row in the UI.

Change the component layout on the Component edit screen.
Change the Group input field on the Component edit screen to a text box.

>  2. Currently in the Project view page in edit mode the Group is editable in the text field and it takes the user organization value from the Liferay DB and from the couchDB, but in the Component edit view the Group is a drop down list. Could you please confirm if this feature for the component view is correct, since it differs from the Project view for the user group.

Change to editable.  The UI requires the Group to be input, but if it is left blank, the primary organization of the User is inserted/.

>3. Private component/release is showing Inaccessible component/release for the other users, so if we have more private component then all will show inaccessible in the UI for other users. But currently in the Projects view, if we set a project visibility to private then it won't be visible to the other users, but in the new changes for component/releases it shows the text as "Inaccessible component"/"Inaccessible release". is this correct?

Change the labels of inaccessible items to "Restricted component"/"Restricted release"

> 4. For other user where the release is Inaccessible for him due to the visibility as Private.
Go to Edit Component->Open a release->Edit Release->linked releases
it shows the Inaccessible release for the user.
Now, if we try to add/link a new release. Here the Inaccessible release text looks bit odd as per the visibility in the UI.

Fixed like above comment and image.
The LinkedRelease edit screen for Release does not show restricted release, but shows the number of releases.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentSearchHandler.java [hunk 1]
```diff
@@ -99,4 +107,23 @@ public ComponentSearchHandler(Supplier<HttpClient> httpClient, Supplier<Cloudant
...
+    }
+
+    public List<Component> searchWithAccessibility(String text, final Map<String , Set<String > > subQueryRestrictions, User user ){
+        List<Component> resultComponentList = connector.searchViewWithRestrictions(Component.class, luceneSearchView, text, subQueryRestrictions);
+        for(Component component : resultComponentList){
+            makePermission(component, user).fillPermissionsInOther(component);
+        }
+        return resultComponentList;
+    }
+
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentSearchHandler.java:128]: The two methods of ComponentSearchHandler were added to control the permissions of Component for both backend and frontend, and have different roles (different outputs), so they are not integrated.

File: libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/permissions/ComponentPermissionsVisibilityTest.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        // @formatter:on
+    }
+
+    @Test
+    @UseDataProvider("componentVisibilityByRoleProvider")
+    public void testVisibilityForComponent(Visibility visibility, String bu, ComponentRole role, String moderatingUser, String creatingUser, String viewingUser, boolean expectedVisibility) throws Exception {
+    	if(role==MODERATOR) {
+    		given().a_component_with_$_$(role, moderatingUser).with_visibility_$_and_business_unit_$(visibility, bu);
+    	}else {
+    		given().a_component_with_$_$(role, creatingUser).with_visibility_$_and_business_unit_$(visibility, bu);
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/permissions/ComponentPermissionsVisibilityTest.java]: fixed

File: libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java [hunk 3]
```diff
@@ -531,6 +558,23 @@ public static String displayNameFor(String name, Map<String, String> nameToDispl
         return releaseNamesMap;
     }
 
+    public static <T> Map<String, T> putAccessibleReleaseNamesInMap(Map<String, T> map, List<Release> releases, User user, T inaccessibleValue) {
+        if(map == null || releases == null) {
+            return Collections.emptyMap();
+        }
+        Map<String, T> releaseNamesMap = new HashMap<>();
+        int inaccessbileReleaseCount = 0;
+        for(Release release : releases) {
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply libraries/lib-datahandler/src/main/java/org/eclipse/sw360/datahandler/common/SW360Utils.java]: fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 4]
```diff
@@ -1641,6 +1686,36 @@ private void removeReleaseId(String releaseId, String componentId) throws SW360E
...
+        for(ReleaseLink releaseLink : releaseLinkList) {
+            Release release = releaseRepository.get(releaseLink.getId());
+            releaseLink.setAccessible(isReleaseActionAllowed(release, user, RequestedAction.READ));
+        }
+        return releaseLinkList;
+    }
+    
+    public boolean isReleaseActionAllowed(Release release, User user, RequestedAction action) {
+        boolean isAllowed = false;
+        switch(action){
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 5]
```diff
@@ -1633,6 +1669,15 @@ private void removeReleaseId(String releaseId, String componentId) throws SW360E
         return getLinkedReleases(project.getReleaseIdToUsage(), visitedIds);
     }
 
+    List<ReleaseLink> getLinkedReleasesWithAccessibility(Project project, Deque<String> visitedIds, User user) {
+        List<ReleaseLink> releaseLinkList = getLinkedReleases(project.getReleaseIdToUsage(), visitedIds);
+        for(ReleaseLink releaseLink : releaseLinkList) {
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 6]
```diff
@@ -304,6 +323,15 @@ public Component getComponent(String id, User user) throws SW360Exception {
         return component;
     }
 
+    public Component getAccessibleComponent(String id, User user) throws SW360Exception {
+        Component component = getComponent(id, user);
+        Map<RequestedAction, Boolean> permissions = component.getPermissions();
+        if(!permissions.get(RequestedAction.READ)) {
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 7]
```diff
@@ -231,10 +231,25 @@ private void autosetReleaseClearingState(Release releaseAfter, Release releaseBe
...
+    }
+
     public List<Release> getRecentReleases() {
         return releaseRepository.getRecentReleases();
     }
-
+    
+    public List<Release> getRecentReleasesWithAccessibility(User user) {
+        List<Release> releaseList = releaseRepository.getRecentReleases();
+        for(Release release : releaseList) {
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: fixed

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 8]
```diff
@@ -1633,6 +1669,15 @@ private void removeReleaseId(String releaseId, String componentId) throws SW360E
         return getLinkedReleases(project.getReleaseIdToUsage(), visitedIds);
     }
 
+    List<ReleaseLink> getLinkedReleasesWithAccessibility(Project project, Deque<String> visitedIds, User user) {
+        List<ReleaseLink> releaseLinkList = getLinkedReleases(project.getReleaseIdToUsage(), visitedIds);
```
[the remark this answers is not in the document]
Comment (KoukiHama) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java:1673]: Modified CommonUtils to perform null checking.

**Comment by KoukiHama** [comment]
> 5. I am not able to see the below highlighted view (Summary subsection (List display of Component in use)), Please let me know if i am missing the step to produce this.

This is the example procedure of how to show this.

1.  Log in as Admin user and create a Component "NewLibrary001".
image

2. Create a Release in Component "NewLibrary001".
image

3. Create Component "NewLibrary002". To verify the Component browsing restriction, set "Group and Moderators" for Component visibility and "DEPARTMENT2" for Group.

image

4. Create a Release for Component "NewLibrary002".

image

5. Create a link from the "NewLibrary002" Release to the "NewLibrary001" Release.

image

6. Log in as a general user who does not belong to DEPARTMENT2, and open the detail screen of Component "NewLibrary001".
image

**Comment by KoukiHama** [comment]
@ravi110336  Thank you for review and comments.
I answered your question and fix codes! Could you check them again?

**Comment by ravi110336** [comment]
Hi @KoukiHama Thanks for incorporating the changes and the clarification, I will have a look into this and give you an update.

**Comment by KoukiHama** [comment]
I solved merge conflict.

**Comment by KoukiHama** [comment]
（from discussion in sw360telco ）This feature needs to be able to be disabled/enabled in sw360.properties.

**Comment by ravi110336** [comment]
Reviewed and tested as per the comments mentioned above, working fine, as discussed in the call after the merge we need to make this feature configurable, for the users who wants to use this feature or not.

**Comment by KoukiHama** [comment]
@ravi110336  Thank you for comment and doing test.

>  we need to make this feature configurable, for the users who wants to use this feature or not.

I have started to realize it!

**Comment by KoukiHama** [comment]
@ravi110336  
I added configure option for Restrict visibility of each component.
However, I am goin to consider test scenario and  do test for configuration on my environment for a while.

**Comment by KoukiHama** [comment]
If anyone have a idea about good test scenario for this function, please let me know.

**Comment by KoukiHama** [comment]
@ag4ums @ravi110336 
Hi, I finally did all test about this pull request on my environment, include configurable function.
Could you review  https://github.com/eclipse/sw360/commit/16f1270e0d651319abf15fbfff6d3ee50362a0f1 again?

if you set **component.visibility.restriction.enabled=false** in sw360.proprty, Restrict feature about component will be off.

**Comment by ravi110336** [comment]
Hi @KoukiHama ,
Thanks for making this config change, I have looked into this new config changes and tested it.
as per the current code by default this feature will be enable for every user until the user set the 
**component.visibility.restriction.enabled=false** in the sw360.properties file to make this feature disabled.

I have discussed this with my team and as per our understanding we want this feature to be disabled by default, and if any user wants to enable this then they can set the sw360.properties file with the below value as true.
**component.visibility.restriction.enabled=true**

Please let me know if i am wrong here.

**Comment by KoukiHama** [comment]
@ravi110336 
Thank you feedback. I will change the default setting.

**Comment by KoukiHama** [comment]
@ravi110336 

I have changed source code, and now "component restriction" is false until user change "component.visibility.restriction.enabled=true".  

Could you check it again?

https://github.com/eclipse/sw360/pull/1357/commits/c9a1bd52d05b90ff774e5ddaef01e0d673f75626

**Comment by KoukiHama** [comment]
Build test error.... I will check it soon.

**Comment by KoukiHama** [comment]
I've added some tests for restricted components. However, if you turn off this feature by default, these tests will be useless.

[libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/permissions/ComponentPermissionsVisibilityTest.java](https://github.com/eclipse/sw360/blob/4b0b26bc74bad480fe297a19a3e16a46dd08ec96/libraries/lib-datahandler/src/test/java/org/eclipse/sw360/datahandler/permissions/ComponentPermissionsVisibilityTest.java)

**Comment by KoukiHama** [comment]
(from discussion in sw360 telco)

Like this couchdb's test code, reading the sw360.property before doing the test for restriction component.

http://github.com/eclipse/sw360/pull/1338/files#diff-21d5cedc14e796af393f061553aca55eaa753b2d6f0365d3bfce93bcbbef7951R121
https://github.com/eclipse/sw360/pull/1338/files#diff-551b1d83d818c3b9f316f9d82cc21760ec6af262af3d518294b63f26e11a388cR79

**Comment by KoukiHama** [comment]
@ravi110336 
Hello, I changed test code and parameters.

> -DRunRestIntegrationTest=true

Now we can mvn install above parameters  and do it in CI by github action yaml.

**Comment by ravi110336** [comment]
Hi @KoukiHama Thanks i will have a look into and give an update.

**Comment by ravi110336** [comment]
Hi @KoukiHama , I have reviewed and tested the PR, code looks good to me and the default setting for this feature as false also works fine. Thanks for incorporating the changes.

Can you please squash the commits to a single commit.

**Comment by KoukiHama** [comment]
@ravi110336 

> Can you please squash the commits to a single commit.

I did it. Could you check it?

**Comment by KoukiHama** [comment]
This "pull request" include migration scripts for updating coucdb, so maybe, need to increase version
https://github.com/eclipse/sw360/blob/681eb0c408d24998a38ca14195695d287e99ad92/scripts/migrations/048_add_component_businessunit.py

**Pull request merged**